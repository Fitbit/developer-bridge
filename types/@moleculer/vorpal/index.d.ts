// Typedoc definitions taken and modified from vorpal source https://github.com/dthree/vorpal

/// <reference types="inquirer" />
declare module '@moleculer/vorpal' {
  import { EventEmitter } from 'events';
  import { Inquirer } from 'inquirer';

  class Vorpal {
    activeCommand: Vorpal.CommandInstance;
    session: Vorpal.Session;
    ui: Vorpal.UiInstance;

    /**
     * Set id for command line history
     * @param id
     * @return {Vorpal}
     * @api public
     */
    history(id: string): this;

    /**
     * Imports a library of Vorpal API commands
     * from another Node module as an extension
     * of Vorpal.
     *
     * @param {Function} commands
     * @return {Vorpal}
     * @api public
     */
    use(commands: (this: this, vorpal: this) => any): this;

    /**
     * Registers a new command in the vorpal API.
     *
     * @param {String} name
     * @param {String} desc
     * @return {Command}
     * @api public
     */
    command(name: string, desc?: string): Vorpal.Command;

    /**
     * Sets the permanent delimiter for this
     * Vorpal server instance.
     *
     * @param {String} str
     * @return {Vorpal}
     * @api public
     */
    delimiter(str: string): this;

    /**
     * Hook the tty prompt to this given instance
     * of vorpal.
     *
     * @return {Vorpal}
     * @api public
     */
    show(): this;

    /**
     * Executes a vorpal API command and
     * returns the response either through a
     * callback or Promise in the absence
     * of a callback.
     *
     * A little black magic here - because
     * we sometimes have to send commands 10
     * miles upstream through 80 other instances
     * of vorpal and we aren't going to send
     * the callback / promise with us on that
     * trip, we store the command, callback,
     * resolve and reject objects (as they apply)
     * in a local vorpal._command variable.
     *
     * When the command eventually comes back
     * downstream, we dig up the callbacks and
     * finally resolve or reject the promise, etc.
     *
     * Lastly, to add some more complexity, we throw
     * command and callbacks into a queue that will
     * be unearthed and sent in due time.
     *
     * @param {String} cmd
     * @param {Object} args
     * @return {Promise}
     * @api public
     */
    exec<T>(
      cmd: string,
      args: { [k: string]: any; sessionId?: string },
    ): Promise<T>;
    exec<T>(cmd: string): Promise<T>;

    /**
     * Executes a Vorpal command in sync.
     *
     * @param {String} cmd
     * @param {Object} args
     * @return {*} stdout
     * @api public
     */
    execSync<T>(cmd: string): T;

    /**
     * Delegates to ui.log.
     *
     * @param {String} log
     * @return {Vorpal}
     * @api public
     */
    log(message?: any, ...optionalParams: any[]): this;

    /**
     * Registers a new 'mode' command in the vorpal API.
     *
     * @param {String} name
     * @param {String} desc
     * @return {Command}
     * @api public
     */
    mode(name: string, desc?: string): Vorpal.Command;
  }

  namespace Vorpal {
    interface Command {
      /**
       * Defines an action for a given command.
       *
       * @param {Function} fn
       * @return {Command}
       * @api public
       */
      action<T>(
        actionFn: (this: CommandInstance, args: Args) => void | Promise<T>,
      ): this;
      action<T>(
        actionFn: (this: CommandInstance, command: string) => void | Promise<T>,
      ): this;

      /**
       * Provides an alias to the command
       *
       * @param {String} alias
       * @return {Command}
       * @api public
       */
      alias(alias: string): this;

      /**
       * Sets args for static typing of options
       * using minimist.
       *
       * @param {Object} types
       * @return {Command}
       * @api public
       */
      types(types: { string: string[] }): this;

      /**
       * Doesn't show command in the help menu.
       *
       * @return {Command}
       * @api public
       */
      hidden(): this;

      /**
       * Defines description for given command.
       *
       * @param {String} str
       * @return {Command}
       * @api public
       */
      description(str: string): this;

      /**
       * Defines a prompt delimiter for a
       * mode once entered.
       *
       * @param {String} delimiter
       * @return {Command}
       * @api public
       */
      delimiter(delimiter: string): this;

      /**
       * Defines an init action for a mode command.
       *
       * @param {Function} fn
       * @return {Command}
       * @api public
       */
      init(fn: () => void | Promise<void>): this;

      /**
       * Registers an option for given command.
       *
       * @param {String} flags
       * @param {String} description
       * @return {Command}
       * @api public
       */
      option(
        flags: string,
        description: string,
        autocomplete?: Autocomplete,
      ): this;

      /**
       * Defines a function to be called when the
       * command is canceled.
       *
       * @param fn
       * @returns {Command}
       * @api public
       */

      cancel(fn: () => void): this;
    }

    type Autocomplete =
      | string[]
      | ((input: string) => string[] | Promise<string>)
      | ((input: string, callback: (completions: string[]) => void) => void);

    interface Args {
      options: {
        [name: string]: string;
      };
    }

    interface CommandInstance {
      log(message?: any, ...optionalParams: any[]): void;
      prompt: Inquirer['prompt'];
    }

    interface Session {
      id: number;
    }

    interface Redraw {
      (...message: string[]): void;
      done(): void;
    }

    interface UiInstance extends EventEmitter {
      redraw: Redraw;
      log(message?: any, ...optionalParams: any[]): void;
    }
  }

  export = Vorpal;
}
