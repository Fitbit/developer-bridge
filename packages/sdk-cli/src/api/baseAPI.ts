import * as t from 'io-ts';

import * as auth from '../auth';
import environment from '../auth/environment';
import fetch, { Headers } from '../fetch';
import { assertJSON } from '../util/fetchUtil';

// tslint:disable-next-line:variable-name
const ErrorResponse = t.type(
  {
    errors: t.array(t.type({
      message: t.string,
    })),
  },
  'Hosts403Response',
);
type ErrorResponse = t.TypeOf<typeof ErrorResponse>;

export const okOrElse = <T>(failure: (response: Response) => Promise<T>) => (response: Response) =>
  response.ok ? Promise.resolve(response) : failure(response);
export const assertOK = okOrElse((response) => {
  return assertJSON()(response)
    .catch(() => Promise.reject(
      // tslint:disable-next-line:max-line-length
      new Error(`Fetch of ${response.url} returned status ${response.status} ${response.statusText}`),
    ))
    .then(responseObject => ErrorResponse.decode(responseObject).getOrElseL(() => {
      // tslint:disable-next-line:max-line-length
      throw new Error(`Fetch of ${response.url} returns status ${response.status} with payload: ${JSON.stringify(responseObject)}`);
    }))
    .then(errorObj => Promise.reject(new Error(errorObj.errors.map(err => err.message).join('\n'))));
});

export async function apiFetch(path: string, init: RequestInit = {}) {
  const authToken = await auth.getAccessToken();
  if (!authToken) throw new Error(`Fetch of ${path} failed: no stored auth token`);

  const headers = new Headers(init.headers || {});
  headers.set('authorization', `Bearer ${authToken}`);
  const mergedInit: RequestInit = {
    cache: 'no-store',
    ...init,
    headers,
  };
  return fetch(`${environment().config.apiUrl}/${path}`, mergedInit).then(assertOK);
}
