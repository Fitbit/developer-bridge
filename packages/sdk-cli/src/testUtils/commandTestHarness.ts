import vorpal from '@moleculer/vorpal';

export default function commandTestHarness(extension: (cli: vorpal) => void) {
  const cli = new vorpal();
  // Quiet the MaxListenersExceeded warning when this harness is used in
  // many tests. Each vorpal instance registers a 'vorpal_ui_keypress'
  // listener on the `ui` singleton, and many vorpal instances are
  // created during a normal test run, so the warning is a false alarm.
  cli.ui.setMaxListeners(0);

  cli.use(extension);

  const mockLog = jest.fn();
  const mockPrompt = jest.fn();

  Object.defineProperty(cli, 'activeCommand', {
    value: {
      log: mockLog,
      prompt: mockPrompt,
    },
  });

  const mockUIRedraw: vorpal.Redraw = Object.assign(jest.fn(), {
    clear: jest.fn(),
    done: jest.fn(),
  });

  cli.ui.redraw = mockUIRedraw;
  cli.ui.log = mockLog;
  cli.log = mockLog;

  return {
    cli,
    mockLog,
    mockPrompt,
    mockUIRedraw,
  };
}
