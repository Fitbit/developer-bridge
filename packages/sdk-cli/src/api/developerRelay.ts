import * as t from 'io-ts';
import { decode } from '@fitbit/jsonrpc-ts';
import stream from 'stream';
import websocketStream from 'websocket-stream';

import { apiFetch } from './baseAPI';
import { assertContentType, assertOK, assertJSON, decodeJSON } from '../util/fetchUtil';

// tslint:disable-next-line:variable-name
export const Host = t.type(
  {
    id: t.string,
    displayName: t.string,
    roles: t.array(t.string),
    state: t.union([
      t.literal('available'),
      t.literal('busy'),
    ]),
  },
  'Host',
);
export type Host = t.TypeOf<typeof Host>;

// tslint:disable-next-line:variable-name
const HostsResponse = t.type(
  {
    hosts: t.array(Host),
  },
  'HostsResponse',
);

// tslint:disable-next-line:variable-name
const Hosts403Response = t.type(
  {
    errors: t.array(t.type({
      message: t.string,
    })),
  },
  'Hosts403Response',
);
type Hosts403Response = t.TypeOf<typeof Hosts403Response>;

async function getConnectionURL(hostID: string) {
  const response = await apiFetch(
    `1/user/-/developer-relay/hosts/${hostID}`,
    { method: 'POST' },
  ).then(assertOK).then(assertContentType('text/uri-list'));
  const uriList = (await response.text())
    .split('\r\n')
    .filter(line => line[0] !== '#');
  return uriList[0];
}

function createWebSocket(uri: string) {
  return new Promise<stream.Duplex>((resolve, reject) => {
    const ws = websocketStream(uri, { objectMode: true });
    ws.on('connect', () => resolve(ws));
    ws.on('error', e => reject(e));
  });
}

export async function connect(hostID: string) {
  const url = await getConnectionURL(hostID);
  return createWebSocket(url);
}

export async function hosts() {
  const response = await apiFetch('1/user/-/developer-relay/hosts.json')
    .then(async (response: Response) => {
      if (response.ok || response.status !== 403) return response;

      const errorObj = await assertJSON()(response).then(decode(Hosts403Response));
      const errorString = errorObj.errors.join('\n');
      throw new Error(errorString);
    })
    .then(decodeJSON(HostsResponse));

  const hostsWithRole = (role: string) =>
    response.hosts.filter(host => host.roles.includes(role));

  return {
    appHost: hostsWithRole('APP_HOST'),
    companionHost: hostsWithRole('COMPANION_HOST'),
  };
}
