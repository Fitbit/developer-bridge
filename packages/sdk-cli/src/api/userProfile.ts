
import * as t from 'io-ts';

import { apiFetch, decodeJSON } from './baseAPI';

// tslint:disable-next-line:variable-name
const UserProfileResponse = t.type(
  {
    user: t.type({
      email: t.string,
      fullName: t.string,
    }),
  },
  'UserProfileResponse',
);

export default async function userProfile() {
  const response = await apiFetch('1/user/-/profile.json')
    .then(decodeJSON(UserProfileResponse));
  return response.user;
}
