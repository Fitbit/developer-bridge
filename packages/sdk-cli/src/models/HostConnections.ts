import 'stream.finished/auto';

import { RemoteHost } from '@fitbit/fdb-debugger';
import dateformat from 'dateformat';
import fs from 'fs';
import stream, { Duplex } from 'stream';
import { SyncEvent } from 'ts-events';

import * as developerRelay from '../api/developerRelay';
import * as USBDebugHost from './USBDebugHost';
import { DeviceType, HostType } from './HostTypes';

import StreamTap from './StreamTap';

export class HostConnection {
  private constructor(public stream: stream.Duplex, public host: RemoteHost) {}

  static getDumpStreamTap() {
    const shouldDumpLogFile = process.env.FITBIT_DEVBRIDGE_DUMP === '1';
    if (!shouldDumpLogFile) return undefined;

    const dumpLogFilePath = dateformat('"log" yyyy-mm-dd "at" H.MM.ss."txt"');
    const dumpLogFileHandle = fs.openSync(dumpLogFilePath, 'w');
    const now = () => new Date().getTime();
    const epoch = now();

    function writeChunk(prefix: string) {
      return (chunk: any) =>
        fs.writeSync(
          dumpLogFileHandle,
          `[${prefix}][${now() - epoch}] ${JSON.stringify(
            chunk,
            undefined,
            2,
          )}\n`,
        );
    }

    const transforms = {
      preSerializeTransform: new StreamTap(writeChunk('send')),
      postDeserializeTransform: new StreamTap(writeChunk('recv')),
    };

    stream.finished(transforms.postDeserializeTransform, () =>
      fs.closeSync(dumpLogFileHandle),
    );

    return transforms;
  }

  static async connect(stream: Duplex) {
    return new this(
      stream,
      await RemoteHost.connect(stream, this.getDumpStreamTap()),
    );
  }
}

export type HostAddedEvent = {
  hostType: HostType;
  host: RemoteHost;
};

export interface Host {
  connect(): Promise<Duplex>;
  displayName: string;
  available: boolean;
  roles: string[];
}

class HostConnections {
  onHostAdded = new SyncEvent<HostAddedEvent>();
  appHost?: HostConnection;
  companionHost?: HostConnection;

  async connect(host: Host, deviceType: DeviceType) {
    const hostTypes: { [key in DeviceType]: HostType } = {
      device: 'appHost',
      phone: 'companionHost',
    };
    const hostType = hostTypes[deviceType];

    const existingHost = this[hostType];
    if (existingHost) existingHost.stream.destroy();

    const stream = await host.connect();

    let hostConnection: HostConnection;
    try {
      hostConnection = await HostConnection.connect(stream);
    } catch (ex) {
      stream.destroy();
      throw ex;
    }

    this[hostType] = hostConnection;
    this.onHostAdded.post({ hostType, host: hostConnection.host });
    return hostConnection;
  }

  async list() {
    const hosts: Host[] = [];

    try {
      for (const device of await developerRelay.hosts()) {
        hosts.push(device);
      }
    } catch (error) {
      throw new Error(
        `An error was encountered when loading the list of available Developer Relay hosts: ${
          (error as Error).message
        }`,
      );
    }

    try {
      for (const device of await USBDebugHost.list()) {
        hosts.push(device);
      }
    } catch (error) {
      throw new Error(
        `An error was encountered when loading the list of available USB hosts: ${
          (error as Error).message
        }`,
      );
    }

    return hosts;
  }

  async listOfType(deviceType: DeviceType) {
    const hosts = await this.list();

    const hostRole = {
      device: 'APP_HOST',
      phone: 'COMPANION_HOST',
    }[deviceType];

    return hosts.filter((host) => host.roles.includes(hostRole));
  }
}

export default HostConnections;
