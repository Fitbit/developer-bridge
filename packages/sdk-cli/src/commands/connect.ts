import { startCase } from 'lodash';
import vorpal from '@moleculer/vorpal';

import * as developerRelay from '../api/developerRelay';
import HostConnections from '../models/HostConnections';

export type DeviceType = 'device' | 'phone';

export const connectAction = async (
  cli: vorpal,
  deviceType: DeviceType,
  hostConnections: HostConnections,
) => {
  let hosts: {
    appHost: developerRelay.Host[];
    companionHost: developerRelay.Host[];
  };

  try {
    hosts = await developerRelay.hosts();
  } catch (error) {
    cli.log(
      // tslint:disable-next-line:max-line-length
      `An error was encountered when loading the list of available ${deviceType} hosts: ${error.message}`,
    );
    return false;
  }

  const hostTypes: { [key: string]: keyof typeof hosts } = {
    device: 'appHost',
    phone: 'companionHost',
  };

  const hostType = hostTypes[deviceType];
  const matchedHosts = hosts[hostType].filter(
    (host) => host.state === 'available',
  );

  if (matchedHosts.length === 0) {
    cli.activeCommand.log(`No ${deviceType}s are connected and available`);
    return false;
  }

  let host: { id: string; displayName: string };
  if (matchedHosts.length === 1) {
    host = matchedHosts[0];
    cli.activeCommand.log(
      `Auto-connecting only known ${deviceType}: ${host.displayName}`,
    );
  } else {
    host = (
      await cli.activeCommand.prompt<{
        hostID: { id: string; displayName: string };
      }>({
        type: 'list',
        name: 'hostID',
        message: `Which ${deviceType} do you wish to sideload to?`,
        choices: matchedHosts.map((host) => ({
          name: host.displayName,
          value: { id: host.id, displayName: host.displayName },
        })),
      })
    ).hostID;
  }

  const connection = await hostConnections.connect(hostType, host.id);
  connection.ws.once('finish', () =>
    cli.log(`${startCase(deviceType)} '${host.displayName}' disconnected`),
  );

  return true;
};

export default function (stores: { hostConnections: HostConnections }) {
  return (cli: vorpal) => {
    const deviceTypes: DeviceType[] = ['device', 'phone'];
    for (const deviceType of deviceTypes) {
      cli
        .command(`connect ${deviceType}`, `Connect a ${deviceType}`)
        .action(async () =>
          connectAction(cli, deviceType, stores.hostConnections),
        );
    }
  };
}
