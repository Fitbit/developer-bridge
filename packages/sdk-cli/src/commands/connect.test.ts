import events from 'events';

import vorpal from 'vorpal';

import connect, { DeviceType } from './connect';
import commandTestHarness from '../testUtils/commandTestHarness';
import * as developerRelay from '../api/developerRelay';
import HostConnections, { HostType } from '../models/HostConnections';

jest.mock('../models/HostConnections');

const mockAppHost: developerRelay.Host = {
  id: 'apphost',
  displayName: 'App Host',
  roles: ['APP_HOST'],
  state: 'available',
};

const mockAppHost2: developerRelay.Host = {
  id: 'apphost2',
  displayName: 'Another App Host',
  roles: ['APP_HOST'],
  state: 'available',
};

const mockCompanionHost: developerRelay.Host = {
  id: 'companionhost',
  displayName: 'Companion Host',
  roles: ['COMPANION_HOST'],
  state: 'available',
};

const mockCompanionHost2: developerRelay.Host = {
  id: 'companionhost2',
  displayName: 'Another Companion Host',
  roles: ['COMPANION_HOST'],
  state: 'available',
};

const mockBusyAppHost: developerRelay.Host = {
  id: 'apphost3',
  displayName: 'Yet Another App Host',
  roles: ['APP_HOST'],
  state: 'busy',
};

const mockRelayHosts = {
  device: [mockAppHost, mockAppHost2],
  phone: [mockCompanionHost, mockCompanionHost2],
};

let cli: vorpal;
let mockLog: jest.Mock;
let mockPrompt: jest.Mock;
let mockWS: events.EventEmitter;

let hostConnections: HostConnections;
let relayHostsSpy: jest.SpyInstance;
let hostConnectSpy: jest.SpyInstance;

const mockRelayHostsResponse = {
  device: (hosts: developerRelay.Host[]) =>
    relayHostsSpy.mockResolvedValueOnce({ appHost: hosts, companionHost: [] }),
  phone: (hosts: developerRelay.Host[]) =>
  relayHostsSpy.mockResolvedValueOnce({ appHost: [], companionHost: hosts }),
};

beforeEach(() => {
  hostConnections = new HostConnections();
  ({ cli, mockLog, mockPrompt } = commandTestHarness(connect({ hostConnections })));
  relayHostsSpy = jest.spyOn(developerRelay, 'hosts');
  hostConnectSpy = jest.spyOn(hostConnections, 'connect');
  mockWS = new events.EventEmitter();
  hostConnectSpy.mockResolvedValueOnce({ ws: mockWS });
});

function doConnect(type: DeviceType) {
  return cli.exec(`connect ${type}`);
}

describe.each([
  ['device', 'appHost'],
  ['phone', 'companionHost'],
])(
  'when the device type argument is %s',
  (deviceType: DeviceType, hostType: HostType) => {
    it(`logs an error if no ${deviceType}s are connected`, async () => {
      mockRelayHostsResponse[deviceType]([]);
      await doConnect(deviceType);
      expect(mockLog.mock.calls[0]).toMatchSnapshot();
    });

    describe(`when a single ${deviceType} is connected`, () => {
      let mockHost: developerRelay.Host;

      beforeEach(() => {
        mockHost = mockRelayHosts[deviceType][0];
        mockRelayHostsResponse[deviceType]([mockHost]);
        return doConnect(deviceType);
      });

      it('does not prompt the user to select a host', () => {
        expect(mockPrompt).not.toBeCalled();
      });

      it('logs a message explaining the auto-connection', () => {
        expect(mockLog.mock.calls[0]).toMatchSnapshot();
      });

      it('acquires a developer relay connection for the given host type and ID', () => {
        expect(hostConnectSpy).toBeCalledWith(hostType, mockHost.id);
      });

      it('logs a message when the host disconnects', () => {
        mockWS.emit('finish');
        expect(mockLog.mock.calls[1]).toMatchSnapshot();
      });
    });

    describe(`when multiple ${deviceType}s are connected`, () => {
      let mockSelectedHost: developerRelay.Host;

      beforeEach(() => {
        const mockHosts = mockRelayHosts[deviceType];
        mockSelectedHost = mockHosts[1];
        mockRelayHostsResponse[deviceType](mockHosts);
        mockPrompt.mockResolvedValueOnce({
          hostID: {
            id: mockSelectedHost.id,
            displayName: mockSelectedHost.displayName,
          },
        });
        return doConnect(deviceType);
      });

      it('prompts the user to select a host', () => {
        expect(mockPrompt).toBeCalled();
      });

      it('acquires a developer relay connection for the given host type and ID', () => {
        expect(hostConnectSpy).toBeCalledWith(hostType, mockSelectedHost.id);
      });
    });

    it('logs an error if the hosts call throws', async () => {
      relayHostsSpy.mockRejectedValueOnce(new Error('some error'));
      await doConnect(deviceType);
      expect(mockLog.mock.calls[0]).toMatchSnapshot();
    });
  },
);

it('does not show busy hosts', async () => {
  mockRelayHostsResponse.device([
    mockAppHost,
    mockAppHost2,
    mockBusyAppHost,
  ]);

  mockPrompt.mockResolvedValueOnce({
    hostID: {
      id: mockAppHost.id,
      displayName: mockAppHost.displayName,
    },
  });

  await doConnect('device');
  expect(mockPrompt).toBeCalledWith(expect.objectContaining({
    choices: [mockAppHost, mockAppHost2].map(host => ({
      value: { id: host.id, displayName: host.displayName },
      name: host.displayName,
    })),
  }));
});

it('does not auto-connect a busy host', async () => {
  mockRelayHostsResponse.device([mockBusyAppHost]);
  await doConnect('device');
  expect(mockLog.mock.calls[0]).toMatchSnapshot();
});
