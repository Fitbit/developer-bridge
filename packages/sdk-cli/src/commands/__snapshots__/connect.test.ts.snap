// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not auto-connect a busy host 1`] = `
Array [
  "No devices are connected and available",
]
`;

exports[`when the device type argument is device logs an error if no devices are connected 1`] = `
Array [
  "No devices are connected and available",
]
`;

exports[`when the device type argument is device logs an error if the hosts call throws 1`] = `
Array [
  "An error was encountered when loading the list of available device hosts: some error",
]
`;

exports[`when the device type argument is device when a single device is connected logs a message explaining the auto-connection 1`] = `
Array [
  "Auto-connecting only known device: App Host",
]
`;

exports[`when the device type argument is device when a single device is connected logs a message when the host disconnects 1`] = `
Array [
  "Device 'App Host' disconnected",
]
`;

exports[`when the device type argument is phone logs an error if no phones are connected 1`] = `
Array [
  "No phones are connected and available",
]
`;

exports[`when the device type argument is phone logs an error if the hosts call throws 1`] = `
Array [
  "An error was encountered when loading the list of available phone hosts: some error",
]
`;

exports[`when the device type argument is phone when a single phone is connected logs a message explaining the auto-connection 1`] = `
Array [
  "Auto-connecting only known phone: Companion Host",
]
`;

exports[`when the device type argument is phone when a single phone is connected logs a message when the host disconnects 1`] = `
Array [
  "Phone 'Companion Host' disconnected",
]
`;
